### DISCUSSION TITLE
Best Motif Analysis Method

### DISCUSSION THREAD

**Discussion by wht10** [stuart-lab/signac#1078] [Q&A] [answered]
Hey @timoast, 

Thanks for making such an amazing tool.

I have noticed that the [motif analysis methods](https://satijalab.org/signac/articles/motif_vignette.html) return fairly different results, and in the case of ChromVar seem to be heavily dependent on the `pwm_set` used. 

I used 3 approaches so far: 
1. ChromVar with JASPAR 2020 pwm
2. ChromVar with pwm supplied by the [greenleaf lab from the csiBP DB](https://github.com/GreenleafLab/chromVARmotifs)
3. Finding Overrepresented motifs using `FindMarkers()` and `FindMotifs()` as outlined in the vignette but with the greenleaf pwm_set to make the `motif.object`

Method 1 only returned 25 motifs, which I suspect is due to the limited size of JASPAR 2020 pwm, and JASPAR 2022 isn't available on BioConductor yet.

Method 2 identified 257 motifs, but the pwm used seems to have some inaccuracies. There are a set of IDs in this pwm that begin with "XP" and the gene name is misattributed to the "LINE ID". 

```r
> head(differential.activity)
                                             p_val avg_diff pct.1 pct.2    p_val_adj     gene
ENSMUSG00000036461-LINE1030-Elf1-I    1.420479e-30 2.290562 0.703 0.108 1.255703e-27     Elf1
XP-9117244-LINE5785-Gm4881-I-N4       7.204168e-27 2.387774 0.724 0.151 6.368485e-24 LINE5785
XP-9117244-LINE5827-Gm4881-I-N1       1.071571e-26 2.098758 0.724 0.194 9.472689e-24 LINE5827
ENSMUSG00000006311-LINE1007-Etv2-I    4.281039e-26 2.097897 0.724 0.161 3.784439e-23     Etv2
ENSMUSG00000040732-LINE1032-Erg-D-N1  1.361230e-24 1.957166 0.677 0.140 1.203328e-21      Erg
ENSMUSG00000032035-LINE1028-Ets1-D-N1 7.540898e-24 1.903893 0.674 0.183 6.666154e-21     Ets1
```
Method 3 identifies 884 motifs.

When I run Method 3 without creating a new `motif.object` then I get a mix of human and murine gene nomenclature, which is also weird. 

```r
head(da_motif)
                                                                      motif observed background percent.observed percent.background
ENSMUSG00000000094_LINE1789_Tbx4_D       ENSMUSG00000000094_LINE1789_Tbx4_D       25       5168         25.51020             10.336
ENSMUSG00000018263_LINE1793_Tbx5_D       ENSMUSG00000018263_LINE1793_Tbx5_D       25       5168         25.51020             10.336
ENSMUSG00000000093_LINE1787_Tbx2_D_N1 ENSMUSG00000000093_LINE1787_Tbx2_D_N1       23       4855         23.46939              9.710
ENSMUSG00000018604_LINE1794_Tbx3_D       ENSMUSG00000018604_LINE1794_Tbx3_D       22       4611         22.44898              9.222
ENSMUSG00000040632_LINE223_Nrl_I           ENSMUSG00000040632_LINE223_Nrl_I       21       4584         21.42857              9.168
ENSMUSG00000047591_LINE639_Mafa_I_N4   ENSMUSG00000047591_LINE639_Mafa_I_N4       21       4584         21.42857              9.168
                                      fold.enrichment       pvalue motif.name
ENSMUSG00000000094_LINE1789_Tbx4_D           2.468093 1.543748e-05       Tbx4
ENSMUSG00000018263_LINE1793_Tbx5_D           2.468093 1.543748e-05       Tbx5
ENSMUSG00000000093_LINE1787_Tbx2_D_N1        2.417033 5.130535e-05       Tbx2
ENSMUSG00000018604_LINE1794_Tbx3_D           2.434285 6.993127e-05       Tbx3
ENSMUSG00000040632_LINE223_Nrl_I             2.337322 1.870158e-04        Nrl
ENSMUSG00000047591_LINE639_Mafa_I_N4         2.337322 1.870158e-04       Mafa
```
```r
 head(da_motif)
            motif observed background percent.observed percent.background fold.enrichment      pvalue    motif.name
MA0006.1 MA0006.1       29       8569         29.59184             17.138        1.726680 0.001614456     Ahr::Arnt
MA0117.2 MA0117.2       19       4928         19.38776              9.856        1.967102 0.003069058          Mafb
MA0002.2 MA0002.2       40      13779         40.81633             27.558        1.481106 0.003114408         RUNX1
MA0160.1 MA0160.1       22       6256         22.44898             12.512        1.794196 0.004327681         NR4A2
MA0503.1 MA0503.1       24       7159         24.48980             14.318        1.710420 0.005162743 Nkx2-5(var.2)
MA0603.1 MA0603.1       10       2563         10.20408              5.126        1.990652 0.028770875         Arntl
```
There doesn't seem to be much consensus in the number of motifs, or the top motifs identified for the same comparison using different methods. I guess my question is, which of these methods seems the most accurate and reliable in your experience? I tried to find some sort of published meta-analysis of the methods, and didn't find anything, but if you could point me in the direction of literature comparing these methods that would be great too. 

Here is the code I used to perform the 3 analyses. 

Method 1
```r
pwm_set <- getMatrixSet(x = JASPAR2020, opts = list(species = 10090, all_versions = FALSE))
motif.matrix <- CreateMotifMatrix(features = granges(gdt17), pwm = pwm_set, genome = 'mm10', use.counts = FALSE)
motif.object <- CreateMotifObject(data = motif.matrix, pwm = pwm_set)
gdt17 <- SetAssayData(gdt17, assay = 'ATAC', slot = 'motifs', new.data = motif.object)

# Note that this step can take 30-60 minutes 
gdt17 <- RunChromVAR(
  object = gdt17,
  genome = BSgenome.Mmusculus.UCSC.mm10
)

#Differential Motif Activity using ChromVar
DefaultAssay(gdt17) <- "chromvar"

differential.activity <- FindMarkers(
  object = gdt17,
  ident.1 = 0,
  ident.2 = 7,
  only.pos = TRUE,
  mean.fxn = rowMeans,
  fc.name = "avg_diff"
)

differential.activity$gene <- ConvertMotifID(gdt17, id = rownames(differential.activity))
```
Method 2
```r
devtools::install_github("GreenleafLab/chromVARmotifs")
library(chromVARmotifs)
data("mouse_pwms_v1")

motif.matrix <- CreateMotifMatrix(features = granges(gdt17), pwm = mouse_pwms_v2, genome = 'mm10', use.counts = FALSE)
motif.object <- CreateMotifObject(data = motif.matrix, pwm = mouse_pwms_v2)
gdt17 <- SetAssayData(gdt17, assay = 'ATAC', slot = 'motifs', new.data = motif.object)

gdt17 <- RunChromVAR(
  object = gdt17,
  genome = BSgenome.Mmusculus.UCSC.mm10
)

#Differential Motif Activity using ChromVar
DefaultAssay(gdt17) <- "chromvar"

differential.activity <- FindMarkers(
  object = gdt17,
  ident.1 = 0,
  ident.2 = 7,
  only.pos = TRUE,
  mean.fxn = rowMeans,
  fc.name = "avg_diff"
)

library(stringr)
motifs <- rownames(differential.activity)
motifs <- str_replace_all(motifs, "-","_")

genes <- sapply(motifs, function(i){
  mouse_pwms_v2[[noquote(i)]]@name
})

differential.activity$gene <- genes

```
Method 3

```r
#Overrepresented motif using Signac
DefaultAssay(gdt17) <- "ATAC"

da_motif <- FindMarkers(
  object = gdt17,
  ident.1 = 0,
  ident.2 = 7,
  only.pos = TRUE,
  test.use = 'LR',
  min.pct = 0.05,
  latent.vars = 'nCount_ATAC'
)

sig_peak <- rownames(da_motif[da_motif$p_val_adj <.05,])

# find peaks open in cluster 0 and 1
open.peaks <- AccessiblePeaks(gdt17, idents = c(0, 7))

# match the overall GC content in the peak set
meta.feature <- GetAssayData(gdt17, assay = "ATAC", slot = "meta.features")
peaks.matched <- MatchRegionStats(
  meta.feature = meta.feature[open.peaks, ],
  query.feature = meta.feature[sig_peak, ],
  n = 50000
)

da_motif <- FindMotifs(gdt17, features = sig_peak, background = peaks.matched)
```

**Comment by timoast** [comment, marked as answer]
The motif collection for individual species in JASPAR is quite limited, I prefer to use the core vertebrate collection which gives 746 motifs.

> I tried to find some sort of published meta-analysis of the methods, and didn't find anything, but if you could point me in the direction of literature comparing these methods that would be great too.

This doesn't exist as far as I know. Basically, the `FindMotifs()` function in Signac is just something I came up with to use a hypergeometric test for enrichment of motifs in a set of peaks, with background peaks matched for GC content.

I prefer to use `FindMarkers()`+`FindMotifs()` rather than relying on chromVAR, and check any enriched motifs using the footprinting functions. This gives you a bit more control because you can make sure you're starting with a set of differential peaks you're confident in (eg, look at coverage plots around the DA peaks), and also optimize the set of background peaks being used in the enrichment test.

**Reply by wht10** [reply]
Thank you @timoast! I guess I will try to stick with the JASPAR Vertebrate collection and the `FindMotifs()` method in Signac. 

You may have already seen this, but it would be cool to maybe add a Motif archetype approach, as an alternative to JASPAR to the Signac documentation. It is in beta, and only available for humans at this point, but it may be an interesting option for the next release of Signac. 

[Motif archetypes](https://twitter.com/jeffvierstra/status/1514973966045048835?s=20&t=TfKlHoUT7M1kWpMXQlYFZA)
[Motif archetype availability in ArchR](https://github.com/GreenleafLab/ArchR/discussions/1386)

**Reply by timoast** [reply]
All we require for building the motif object is a list of position weight or frequency matrices, so users are not limited to what's in JASPAR and can already use the Vierstra archetypes if they want to

**Reply by wht10** [reply]
That is great!

Would [this pwm](https://resources.altius.org/~jvierstra/projects/motif-clustering-v2.0beta/consensus_pwms.meme) work with mouse data, similar to how the vertebrate JASPAR collection doesn't exclusively contain mouse motifs?

Also, do you have any tips on how to convert the linked .txt file to a PWMatrixList object?

**Reply by timoast** [reply]
> Would [this pwm](https://resources.altius.org/~jvierstra/projects/motif-clustering-v2.0beta/consensus_pwms.meme) work with mouse data, similar to how the vertebrate JASPAR collection doesn't exclusively contain mouse motifs?

I don't know, I haven't looked at the Vierstra motif archetypes

> Also, do you have any tips on how to convert the linked .txt file to a PWMatrixList object?

It looks like MEME format, so you could try using the universalmotif package: https://bioconductor.org/packages/release/bioc/html/universalmotif.html

**Reply by wht10** [reply]
Awesome, thanks. 

I have just a couple more questions. 

1. Is the rationale behind using all JASPAR vertebrate motifs, on a mouse dataset that these motifs are highly evolutionarily conserved or is there a different rationale for why it is ok to use non-mouse motifs on a mouse dataset?
2. Also, from what I can tell, the `FindMotifs()` function provides uncorrected p-values. Is multiple testing correction appropriate here, and if so, is there a method you can recommend?

**Reply by timoast** [reply]
The TF binding motifs are highly conserved across species, so you can use the JASPAR vertebrate collection for mouse analysis.

You should correct for multiple testing, and actually I don't know why we aren't providing that already in the output. I've added this now to the develop branch, and you can set the correction method using the `p.adjust.method` parameter. I usually go with BH and have set that as the default.
